src/site/pages/digital/src/containers/ExprToCircuitPopup/generate.ts

```diff
@@ -1,73 +1,154 @@
     }
 
     // Create output LED
-    const o = new LED();
-    o.setName("Output");
+    const o = Create<DigitalComponent>(options.output);
+    action.add(new SetNameAction(o, "Output").execute());
 
     // Get the generated circuit
-    const circuit = ExpressionToCircuit(inputMap, expression, o, ops);
+    let circuit = new DigitalObjectSet();
+    try {
+        circuit = ExpressionToCircuit(inputMap, expression, o, ops);
+    } catch (err) {
+        action.undo(); // Undo any actions that have been done so far
+        throw err;
+    }
+
+    action.add(new AddGroupAction(info.designer, circuit).execute());
 
     // Get the location of the top left corner of the screen, the 1.5 acts as a modifier
     //  so that the components are not literally in the uppermost leftmost corner
     const startPos = info.camera.getPos().sub(info.camera.getCenter().scale(info.camera.getZoom()/1.5));
-    const action = new GroupAction([CreateDeselectAllAction(info.selections).execute(),
-                                    new AddGroupAction(info.designer, circuit).execute()]);
-    const [negateAction, negated] = CreateNegatedGatesAction(info.designer, circuit);
-    action.add(negateAction);
-    OrganizeMinDepth(negated, startPos);
-
-    if (isIC) { // If creating as IC
-        const data = ICData.Create(negated);
+    // TODO: Replace with a better (action based) way of organizing a circuit
+    OrganizeMinDepth(circuit, startPos);
+
+    const circuitComponents = circuit.getComponents();
+
+    // Add labels if necessary
+    if (options.label)
+        addLabels(inputMap, action, circuitComponents, info.designer);
+
+    // Set clock frequencies, also connect to oscilloscope if that option is set
+    if (options.input === "Clock")
+        setClocks(inputMap as Map<string, Clock>, action, options.connectClocksToOscope, o, info.designer);
+
+    if (options.isIC) { // If creating as IC
+        const data = ICData.Create(circuitComponents);
         if (!data)
             throw new Error("Failed to create ICData");
         data.setName(expression);
         const ic = new IC(data);
-        ic.setName(expression);
-        ic.setPos(info.camera.getPos());
-        action.add(CreateDeleteGroupAction(info.designer, negated.getComponents()).execute());
+        action.add(new SetNameAction(ic, expression).execute());
         action.add(new CreateICDataAction(data, info.designer).execute());
+        action.add(CreateDeleteGroupAction(info.designer, circuitComponents).execute());
         action.add(new PlaceAction(info.designer, ic).execute());
+        action.add(new TranslateAction([ic], [ic.getPos()], [info.camera.getPos()]).execute());
         action.add(new SelectAction(info.selections, ic).execute());
     } else { // If placing directly
-        action.add(CreateGroupSelectAction(info.selections, negated.getComponents()).execute());
+        action.add(CreateGroupSelectAction(info.selections, circuitComponents).execute());
     }
```

Comment:
I think also add this one to its own function

And also I think you can just pass in `options` to each function and they can check themselves for `if (!options.isIC) return false;` for example